include/exadg/postprocessor/time_control.cpp

```diff
@@ -0,0 +1,173 @@
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+// C/C++
+#include <limits>
+
+// deal.II
+#include <deal.II/base/exceptions.h>
+
+// ExaDG
+#include <exadg/postprocessor/time_control.h>
+#include <exadg/utilities/print_functions.h>
+
+namespace ExaDG
+{
+TimeControlData::EvalType
+get_unsteady_evaluation_type(TimeControlData const & data)
+{
+  bool const interval = (data.trigger_interval > 0.0);
+  bool const timestep = Utilities::is_valid_timestep(data.trigger_every_time_steps);
+
+  // only one of both can be true
+  if(interval && timestep)
+    return TimeControlData::EvalType::Invalid;
+
+  if(interval)
+    return TimeControlData::EvalType::Interval;
+
+  if(timestep)
+    return TimeControlData::EvalType::Timestep;
+
+  return TimeControlData::EvalType::Invalid;
+}
+
+TimeControlData::TimeControlData()
+  : is_active(false),
+    start_time(std::numeric_limits<double>::max()),
+    end_time(std::numeric_limits<double>::max()),
+    trigger_interval(-1.0),
+    trigger_every_time_steps(numbers::invalid_timestep)
+{
+}
+
+void
+TimeControlData::print(dealii::ConditionalOStream & pcout, bool const unsteady) const
+{
+  if(unsteady || is_active)
+  {
+    print_parameter(pcout, "TimeControl start time", start_time);
+    print_parameter(pcout, "TimeControl end_time", end_time);
+    if(get_unsteady_evaluation_type(*this) == TimeControlData::EvalType::Interval)
+      print_parameter(pcout, "TimeControl triggers every interval", trigger_interval);
+    if(get_unsteady_evaluation_type(*this) == TimeControlData::EvalType::Timestep)
+      print_parameter(pcout, "TimeControl triggers every timestep", trigger_every_time_steps);
+  }
+}
+
+TimeControl::TimeControl()
+  : EPSILON(1.0e-10), reset_counter(true), counter(0), end_time_reached(false)
+{
+}
+
+void
+TimeControl::setup(TimeControlData const & time_control_data_in)
+{
+  time_control_data = time_control_data_in;
+}
+
+// If setup() is not called needs_evaluation() will always evaluate to false
+bool
+TimeControl::needs_evaluation(double const time, types::time_step const time_step_number) const
+{
+  if(time_control_data.is_active == false)
+    return false;
+
+  // steady case
+  if(!Utilities::is_unsteady_timestep(time_step_number))
+  {
+    ++counter;
+    return true;
+  }
+
+  // unsteady evaluation
+  AssertThrow(get_unsteady_evaluation_type(time_control_data) != TimeControlData::EvalType::Invalid,
+              dealii::ExcMessage(
+                "Unsteady evaluation not possible since TimeControlData::EvalType::Invalid"));
```

Comment:
This is hard to interpret for a user, since a user does not know what ``TimeControlData::EvalType::Invalid`` is internally. The question is do we need the assert here or would it be better to place the assert in `get_unsteady_evaluation_type()`? Following this strategy, no comment would be needed in `get_unsteady_evaluation_type()`, since the assert message already documents the logic. Further, we could get rid of `TimeControlData::EvalType::Invalid`, since we call below `get_unsteady_evaluation_type()`, automatically invoking the exception in case the parameters are invalid.

Reply:
The assert can not be in ```get_unsteady_evaluation_type()``` since it is OK to have ```EvalType::Invalid``` if ```is_active=false``` or if it is a steady simulation. I will rename the parameters and change the assert such that it is more clear.

Reply:
I understand the implementation like this: `get_unsteady_evaluation_type()` is never called in case of `Is_active = false` or in case of a steady simulation. Therefore, could you point me to the place in the code where this happens?

Reply:
I currently do not have access to a Computer. But you can look e.g. for the Assert in the setup() function of PointwiseOutput generator if I recall correctly.

Reply:
I looked at pointwise output generator, but do not see a conflict here if `get_unsteady_evaluation_type()` cared about the assert.

Reply:
https://github.com/exadg/exadg/pull/262/files#r1008219228